pkg/cli/admin/ocpcertificates/certgraphanalysis/certkeypair.go

```diff
@@ -0,0 +1,213 @@
+package certgraphanalysis
+
+import (
+	"crypto/ecdsa"
+	"crypto/rsa"
+	"crypto/x509"
+	"fmt"
+	"os"
+
+	"github.com/openshift/oc/pkg/cli/admin/ocpcertificates/certgraphapi"
+	"k8s.io/apimachinery/pkg/util/duration"
+)
+
+func toCertKeyPair(certificate *x509.Certificate) (*certgraphapi.CertKeyPair, error) {
+	ret := &certgraphapi.CertKeyPair{
+		Name: fmt.Sprintf("%v::%v", certificate.Subject.CommonName, certificate.SerialNumber),
+		Spec: certgraphapi.CertKeyPairSpec{
+			SecretLocations: nil,
+			OnDiskLocations: nil,
+			CertMetadata:    toCertKeyMetadata(certificate),
+		},
+	}
+
+	details, err := toCertKeyPairDetails(certificate)
+	ret.Spec.Details = details
+	if err != nil {
+		ret.Status.Errors = append(ret.Status.Errors, err.Error())
+	}
+
+	return ret, nil
+}
+
+func toCertKeyPairDetails(certificate *x509.Certificate) (certgraphapi.CertKeyPairDetails, error) {
+	count := 0
+	isClient := false
+	isServing := false
+	isSigner := (certificate.KeyUsage & x509.KeyUsageCertSign) != 0
+	if isSigner {
+		count++
+	}
+	for _, curr := range certificate.ExtKeyUsage {
+		if curr == x509.ExtKeyUsageClientAuth {
+			isClient = true
+			count++
+		}
+		if curr == x509.ExtKeyUsageServerAuth {
+			isServing = true
+			count++
+		}
+	}
+	var typeError error
+	if count == 0 {
+		typeError = fmt.Errorf("you have a cert for nothing?")
+	}
+	if count > 1 {
+		typeError = fmt.Errorf("you have a cert for more than one?  We don't do that. :(")
+	}
+
+	ret := certgraphapi.CertKeyPairDetails{}
+	if isClient {
+		ret.CertType = "ClientCertDetails"
+		ret.ClientCertDetails = toClientCertDetails(certificate)
+	}
+	if isServing {
+		ret.CertType = "ServingCertDetails"
+		ret.ServingCertDetails = toServingCertDetails(certificate)
+	}
+	if isSigner {
+		ret.CertType = "SignerCertDetails"
+		ret.SignerDetails = toSignerDetails(certificate)
+	}
+
+	if count > 1 {
+		ret.CertType = "Multiple"
+	}
+
+	return ret, typeError
+}
+
+func toClientCertDetails(certificate *x509.Certificate) *certgraphapi.ClientCertDetails {
+	return &certgraphapi.ClientCertDetails{
+		Organizations: certificate.Subject.Organization,
+	}
+}
+
+func toServingCertDetails(certificate *x509.Certificate) *certgraphapi.ServingCertDetails {
+	ret := &certgraphapi.ServingCertDetails{
+		DNSNames:    certificate.DNSNames,
+		IPAddresses: nil,
+	}
+
+	for _, curr := range certificate.IPAddresses {
+		ret.IPAddresses = append(ret.IPAddresses, curr.String())
+	}
+
+	return ret
+}
+
+func toSignerDetails(certificate *x509.Certificate) *certgraphapi.SignerCertDetails {
+	return &certgraphapi.SignerCertDetails{}
+}
+
+func toCertKeyMetadata(certificate *x509.Certificate) certgraphapi.CertKeyMetadata {
+	ret := certgraphapi.CertKeyMetadata{
+		CertIdentifier: certgraphapi.CertIdentifier{
```

Comment:
a proper identifier would consist of all of:
Subject
Issuer
SerialNumber

Reply:
Subject versus CommonName?  Refactor welcome after vendoring into cluster-debug-tools and seeing outcomes.